src/test/java/oripa/domain/cptool/LineAdderTest.java

```diff
@@ -34,40 +37,131 @@
+		var line = new OriLine(200, 0, -200, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(4, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(-200, -200, 0, 0, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 200, 200, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(200, 0, 0, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, -200, 0, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_remove_duplicated_segments_if_two_lines_of_same_type_are_overlapping() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(new OriLine(0, 0, 200, 0, OriLine.Type.VALLEY)));
+		var line = new OriLine(0, 0, 100, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(2, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 100, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(100, 0, 200, 0, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_split_existing_lines_if_new_line_touches_them() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(DIAGONAL_LINE));
+		var line = new OriLine(200, 0, 0, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(3, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(-200, -200, 0, 0, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 200, 200, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(200, 0, 0, 0, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_replace_overlapping_parts() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(new OriLine(-200, -200, 0, 0, OriLine.Type.MOUNTAIN)));
+		var line = new OriLine(-100, -100, 100, 100, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(3, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(-200, -200, -100, -100, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(-100, -100, 0, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 100, 100, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_remove_multiple_fully_overlapping_line_with_splits() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(
+				new OriLine[] { new OriLine(-100, 0, -50, 0, OriLine.Type.VALLEY),
+						new OriLine(0, 0, 50, 0, OriLine.Type.VALLEY),
+						new OriLine(100, 0, 150, 0, OriLine.Type.VALLEY) }));
+
+		var line = new OriLine(-200, 0, 200, 0, OriLine.Type.MOUNTAIN);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(7, creasePattern.size());
+
+		var lastX = -200;
+		for (int x = -100; x <= 150; x += 50) {
+			assertTrue(creasePattern.contains(new OriLine(lastX, 0, x, 0, OriLine.Type.MOUNTAIN)));
```

Comment:
`Collection::contains` doesn't care epsilon of `double`. That might cause a failure of this test if the implementation of `LineAdder::add` is correct.

Reply:
That is a good find, I was trying to fix the failing test, but somehow missed the that there is no epsilon taken into account for the `OriLine.equals` function (or rather `OriPoint.compareTo` since that is used for comparing lines). That should be the case though, right?

Reply:
`LineAdder::addLine` was faulty.
lifesbest23@66effa7321b07a0ada28ce818f833ec80453bb3a should resolve this.